### PR TITLE
Add a couple of symbol/label API features to map renderer jobs 

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR adds a pair of API features (developed in parallel as both needed for a forthcoming processing algorithm):
- a mechanism to plug in a label sink as part of a map renderer job
- a mechanism to skip symbol rendering _while still processing labels_ during a map renderer job

The algorithm which will come in a separate PR aims at exporting label details (x/y position, font family, weight, etc.) for a given map theme, extent, and scale.

The label sink part is interesting as it open the doors to all sorts of cool applications. For e.g., QML apps could prefer rendering labels as QML items instead of drawn on the map canvas.

@nyalldawson , was there a reason why QgsLabelSink wasn't exposed in the python API?

### CODE REVIEW

File: src/analysis/processing/qgsalgorithmvectorize.cpp [hunk 1]
```diff
@@ -113,7 +113,7 @@ QVariantMap QgsVectorizeAlgorithmBase::processAlgorithm( const QVariantMap &para
       for ( int column = 0; column < iterCols; column++ )
       {
         const double value = rasterBlock->valueAndNoData( row, column, isNoData );
-        if ( !isNoData )
+        if ( !isNoData && !qgsDoubleNear( value, 0.0 ) )
```
Comment (nyalldawson) [inline src/analysis/processing/qgsalgorithmvectorize.cpp:116]: What's this for? Looks unrelated

File: src/core/maprenderer/qgsmaprendererjob.h [hunk 2]
```diff
@@ -352,6 +352,22 @@ class CORE_EXPORT QgsMapRendererJob : public QObject SIP_ABSTRACT
...
+     * Returns the label sink associated to this rendering job.
+     * \note not available in Python bindings
+     * \since QGIS 3.24
+     */
+    QgsLabelSink *labelSink() const { return mLabelSink; } SIP_SKIP
+
+    /**
+     * Assigns the label sink which will take over responsability for handling labels
+     * during the rendering job.
+     * \note the ownership is not transferred
```
Comment (nyalldawson) [inline src/core/maprenderer/qgsmaprendererjob.h:365]: Ownership is not transferred and the sink must exist for the lifetime of the map rendering job.

File: src/core/qgsrendercontext.h [hunk 3]
```diff
@@ -531,11 +538,18 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
...
     /**
-     * Assign new labeling engine
+     * Assigns the labeling engine
      * \note not available in Python bindings
      */
     void setLabelingEngine( QgsLabelingEngine *engine ) { mLabelingEngine = engine; } SIP_SKIP
 
+    /**
+     * Assigns the label sink which will take over responsability for handling labels.
+     * \note the ownership is not transferred
```
Comment (nyalldawson) [inline src/core/qgsrendercontext.h:548]: ...and the sink must exist for the lifetime of the render context

File: src/core/qgsrendercontext.h [hunk 4]
```diff
@@ -1024,9 +1038,12 @@ class CORE_EXPORT QgsRenderContext : public QgsTemporalRangeObject
 
     double mSymbologyReferenceScale = -1;
 
-    //! Newer labeling engine implementation (can be NULLPTR)
+    //! Labeling engine implementation (can be NULLPTR)
     QgsLabelingEngine *mLabelingEngine = nullptr;
 
+    //! Label sink (can be NULLPTR)
+    QgsLabelSink *mLabelSink = nullptr;
```
Comment (nyalldawson) [inline src/core/qgsrendercontext.h:1045]: Need to update the copy and assign constructors to copy this member too

File: src/core/vector/qgsvectorlayerrenderer.cpp [hunk 5]
```diff
@@ -713,14 +714,35 @@ void QgsVectorLayerRenderer::prepareLabeling( QgsVectorLayer *layer, QSet<QStrin
...
         {
           engine2->removeProvider( mLabelProvider );
-          mLabelProvider = nullptr; // deleted by engine
+          mLabelProvider = nullptr;
+        }
+      }
+      else
+      {
+        mLabelProvider = layer->labeling()->provider( layer );
+        if ( mLabelProvider )
```
Comment (nyalldawson) [inline src/core/vector/qgsvectorlayerrenderer.cpp:749]: This common code could sit outside the `if ( context.labelSink() )` condition

File: src/analysis/processing/qgsalgorithmvectorize.cpp (hunk 1 above)
Comment (nirvn) [reply src/analysis/processing/qgsalgorithmvectorize.cpp:116]: It is unrelated, I'll talk to you about that one later :wink:

**Comment by nirvn** [comment]
@nyalldawson , thanks, review addressed.

**Review by nyalldawson — APPROVED** [review]
